### PR TITLE
Add error handling to C/Go bindings

### PR DESCRIPTION
**Pull request by haaawk**

### CODE REVIEW

**Comment by psarna** [comment]
I think there's one fundamental issue with returning an error struct allocated on heap - not handling errors will immediately cause a leak. And I think that there are legit cases for not handling errors sometimes, by doing some best effort optimizations, e.g.
```c
    // it's nice if it works, but if it fails e.g. due to SQLITE_BUSY, we just proceed
    libsql_execute("VACUUM");
    // we just leaked the allocated error by ignoring the return value from libsql_execute
    (continue with program logic)
```
. And that's mostly problematic because of the way SQLite works with `SQLITE_BUSY` - it's often an expected error, that is also not fatal, and perfectly reasonable to ignore in some cases.

Another minor issue is that if something errored out due to out-of-memory, allocating the error string will fail as well.

Customarily, SQLite API just returns error codes that you can translate to strings with `sqlite3_errstr`, and if the error message really is dynamic, a function takes a `char *` parameter which is a buffer for the error message -- or NULL, if the caller decides that the error message is not interesting and should not be returned.

So, to sum up, I think a clearer approach would be to return an error code wherever applicable, and return custom error messages by taking a `char *errmsg` parameter, so that we don't allocate on error, and there's no need to remember to call error_free on failure - e.g. in cases where the error is consciously ignored.

**Comment by haaawk** [comment]
I'm not sure it's a good idea to optimise the API for "wrong" usages that ignore returned values. It is a normal thing to expect that a caller is responsible for pointers it receives from the API and making API less convenient just to protect users who don't use the API correctly is not necessarily a good thing. I thought that we want to create an API that's nicer to use than dated SQLite API, no? Using output buffers is error prone - what if the buffer is not big enough?

**Comment by haaawk** [comment]
Finally, ignoring error correctly is not really hard with the current approach:
```
libsql_free_error(libsql_execute("VACUUM"))
```

**Comment by haaawk** [comment]
I implemented the middle ground we discussed offline @psarna. Error message is now an output parameter and if null it's not populated.

**Comment by penberg** [comment]
I am merging to move forward, but I think a DuckDB-like `https://duckdb.org/docs/api/c/api.html#duckdb_result_error` makes more sense than having every function take a parameter for _error messages_.
